### ISSUE TITLE
Create new image classification experiments w/ GPUs

### ISSUE THREAD

**Issue by stsievert** [stsievert/adadamp-experiments#1]
It'd be nice to test out the Dask implementation of batch size growth to see how the training time is affected.

I'm thinking that three damping techniques should be used:

1. Decreases the learning rate
2. Increases the batch size
3. Decreasing the learning rate initially, then increasing the batch size

@JoeHolt it'd be great if you could implement this experiment.

**Comment by stsievert** [comment]
I think the experiments in "[Don't decay the learning rate, increase the batch size][1]" should be mirrored, specifically in their Section 5.1. I'm looking to mirror their figures 2 and 3.

* Model: [Wide-ResNet]
    * Architecture: in the "16-4" architecture (`depth=16`, `widen_factor=4`)
    * "Ghost batch size" of 128.
    * Dropout: 30% probability.
* Dataset: [CIFAR10] (from torchvision)
* Loss: [cross-entropy][2]
* Optimizer: [SGD]
    * Learning rate: 0.1
    * Momentum: 0.9 (w/ Nesterov momentum)
    * Weight decay of 0.5e-3.

The training process will be run for 200 epochs. The same damping schedule will be used for all optimizers, one that increases the "damping factor" by a factor of 5 every 60 epochs.

I think this configuration should work:

``` python
model = DaskClassifier(
    module=WideResNet
    module__depth=16,
    module__width=4,
    module__dropout_rate=0.3,
    loss=nn.CrossEntropyLoss,
    optimizer=optim.SGD,
    optimizer__lr=0.1,
    optimizer__momentum=0.9,
    optimizer__nesterov=True,
    optimizer__weight_decay=0.5e-3,
    batch_size=128,
)
```

Please double-check this. This might require modifying `DaskClassifier.__init__` to accept more keyword arguments. And of course we'll need to add classes that implement variable batch sizes.

[2]:https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html
[CIFAR10]:https://pytorch.org/docs/stable/torchvision/datasets.html#torchvision.datasets.CIFAR10
[Wide-ResNet]:https://github.com/meliketoy/wide-resnet.pytorch/blob/master/networks/wide_resnet.py
[1]:https://arxiv.org/abs/1711.00489
[SGD]:https://pytorch.org/docs/stable/optim.html#torch.optim.SGD

**Comment by stsievert** [comment]
I think it'd be good to vary two parameters:

* Batch size increase/learning rate decay schedule
* Maximum number of workers.

In Sec. 5.1, Smith et al. run SGD with these batch sizes:

* [128, 640, 3200, 16000].
* [128, 640].
* [128].

The batch sizes increases by a factor of 5 every 60 epochs. If a maximum batch size is reached, the learning rate decays by a factor of 5 instead.

I think it'd be interesting to see batch sizes of [128, 640, 3200] in addition. I think it'd be reasonable to have each worker process 128 or 640 gradients; 128 would correspond to [1, 5, 25, 125] workers.

I'd run two sets of experiments, with the following goals:

1. Collect `gradient` timings with different batch sizes. The timings with batch sizes 128 and 640 will be most relevant.
2. Collect loss information when varying the batch size/learning rate with the above schedules.

(2) might require mirroring [damping.py#L201-L205][2] in `_dist.py#gradient` ([_dist.py#L76-L79][1]). With this data, it will be possible to simulate the experiments with the saved history.

<details>
<summary>Draft implementation of <code>ReplayHistory</code></summary>

``` python
from typing import List

# class ReplayHistory(BaseEstimator)
def __init__(self, history: List[dict], random_state: int=0):
    """
    history is a list of score history:
    >>> history[0].keys() == {"test_accuracy", "test_loss", "train_accuracy", "train_loss",
                              "score_calls", "partial_fit_calls", "num_data", "model_updates", ...}
    """
    self.history = history
    self.gradient_timings = {128: [0.1, 0.11, 0.09], 128*5: [0.5, 0.51, 0.49]}
    self.random_state = random_state

def initialize(self):
    self.random_state_ = check_random_state(self.random_state)
    self._pf_calls = 0
    self._score_calls = 0

def partial_fit(self, X, y=None):
    self._pf_calls += 1
    rng = self.random_state_
    n_workers = self.batch_size() // 128
    worker_bs = self.batch_size() // n_workers
    batch_timings = self.gradient_timings[worker_bs]
    worker_sleep_times = [rng.sample(batch_timings) for _ in range(n_workers)]
    sleep(max(worker_sleep_times))
    return self

def score(self, X, y=None, prefix="test"):
    self._score_calls += 1
    datum: dict = _find_history_timing(self.history, self._score_calls, self._pf_calls)
    return datum[f"{prefix}_accuracy"]
```

</details>


[2]:https://github.com/stsievert/adadamp/blob/4302e836c35e900245ca4bdd29611866ff96613e/adadamp/damping.py#L201-L205
 [1]:https://github.com/stsievert/adadamp/blob/4302e836c35e900245ca4bdd29611866ff96613e/adadamp/_dist.py#L76-L79

**Comment by JoeHolt** [comment]
Okay to make sure i am on right page:

When would the batch size change? Also what do you mean when you say " 128 would correspond to [1, 5, 25, 125] workers."?

Also is the number of works being varied done purely through the dask client?

Give each worker 128 gradients, so it will scale with batch size in all experiments 


Update:
Experiment 1: 
- Run wide resnet, start decaying learning size right away
- 60 epochs at each size

Experiment 2:
- increase batch size for 120 epochs, then start decreasing learning rate
- 128 for first 60, 640 for second 60, for third 60 do learning rate
- 1 worker initially, then 5

Experiment 3:
- 128 batch size whole way, 0.1 for first 60, then 0.2 for second 60 then n / 5 for last 60
- 1 worker all the way through 

Fix client issue:
- Make dask schedueler in new terminal
- Also make new dask worker in new terminal
- Connect to this client from ipynotebook

**Comment by stsievert** [comment]
I would set a static batch size and make sure that's working first. I'd use the command `nvidia-smi`, which is NVIDIA's version of `htop` and shows the memory/CPU usage of the GPU for each process.

The batch size increase by a factor of 5 every 60 epochs. If a maximum batch size is reached, the learning rate decays by a factor of 5 instead (https://github.com/stsievert/adadamp-experiments/issues/1#issuecomment-745452312 is edited).

**Comment by JoeHolt** [comment]
Why was the meta_ changed to be a singular dictionary, rather than an array of dictionaries? It seems like with the current sets up you cant track states using the meta object, you can only get the meta info from the last epoch ran.

**Comment by stsievert** [comment]
@JoeHolt could you make a PR with your branch? https://github.com/stsievert/adadamp-experiments/compare/dask-exp?expand=1

**Comment by JoeHolt** [comment]
> Thanks for this implementation @JoeHolt! Most of my comments are style/architecture nits. I've created #16 with some other small fixes.
> 
> I'll reply to [stsievert/adadamp-experiments#1 (comment)](https://github.com/stsievert/adadamp-experiments/issues/1#issuecomment-748688115) here; this seems like a better place.
> 
> > Why was the meta_ changed to be a singular dictionary, rather than an array of dictionaries? It seems like with the current sets up you cant track states using the meta object, you can only get the meta info from the last epoch ran.
> > —@JoeHolt
> 
> That's exactly the point – I only want this library to record information necessary to train machine learning models. This is more an architecture nit than anything. If the user of this library wants to record the history every epoch, they can do that (that's all the previous implementation did IIRC). Here's a workflow that I typically use:
> 
> ```python
> est = DaskBaseDamper(...)
> X_train, y_test = ...
> X_test, y_test = ...
> 
> history = []
> for epoch in range(10):
>     est.partial_fit(X_train, y_train)
>     score = est.score(X_test, y_test)
>     datum = {"epoch": epoch, "score": score, **est.get_params(), **est.meta_}
>     history.append(datum)
> 
> import pandas as pd
> pd.DataFrame(history).to_csv("history.csv")
> ```
> 
> This means that the user will essentially opt-in to history recording, and this library has a minimal computational/memory footprint.



I reworked how the training notebook works based on your comment. The train function can now be passed different schedules for the bs and learning rate. Code for this is in the previous PR. Should i make a different PR for this?

**Comment by stsievert** [comment]
Hm... in Section 5.2, Smith et al. describe some modifications on their experiment that reduce the number of model updates. They (almost) use these parameters:

* `lr *= 5, initial_batch_size=640`
* `lr *= 5, initial_batch_size=640, momentum=0.98`

Smith et al. sets `max_batch_size=5120`. I'd use that.

Could you run experiments with those parameters @JoeHolt? I'm curious to see if the number of updates is reduced.

**Comment by JoeHolt** [comment]
Sounds good.

Running the first experiment now with the following schedules: 
epochs = [0, 60, 120, 180]
lr = [
    (0.05 * 5), 
    (0.05 * 5), 
    (0.05 * 5), 
    (0.05 * 5)
]
bs = [640, 3200, 5120, 5120]

**Comment by stsievert** [comment]
That all looks right except for the definition of `bs`. I think `bs = [640, 640 * 5, 640 * 5 * 5, 640 * 5 * 5 * 5]` should be used. I think `DaskClassifier(..., max_batch_size=5120)` should be used (IIRC `BaseDaskDamper takes a parameters `max_batch_size`).

**Comment by JoeHolt** [comment]
Sounds good, I can switch to using that method. Will that be functionally equivalent?

Also, was that max BS of ~5000 used in the other experiments? I'm worried that may have impacted something in the runs where BS got up 16,000.

**Comment by stsievert** [comment]
> Also, was that max BS of ~5000 used in the other experiments? I'm worried that may have impacted something in the runs where BS got up 16,000.

Their first experiment is replicated successfully, and the maximum batch size is 16,000 in the experiments.

> Will that be functionally equivalent?

No, the two methods are not equivalent. Section 5.1 is focused on decaying the variance of the model update. The variance is proportional to `lr / bs`. This values should become smaller by a factor of 5 at epochs 60, 120 and 180 (either by increasing the batch size or decreasing the learning rate). Setting `max_batch_size` manually will not change this value.
